octobot_commons/async_job.py

```diff
@@ -0,0 +1,202 @@
+
+    async def _run_periodic_task(self, **kwargs):
+        """
+        Calls _run() periodically until self.should_stop == True or cancellation
+        """
+        while not self.should_stop:
+            self.is_started = True
+            elapsed_time_since_last_execution = time.time() - self.last_execution_time
+            if elapsed_time_since_last_execution < self.min_execution_delay:
+                await asyncio.sleep(self.min_execution_delay)
+            else:
+                await asyncio.sleep(
+                    0
+                    if elapsed_time_since_last_execution
+                    >= self.execution_interval_delay
+                    else self.execution_interval_delay
+                )
+            await self._run_task_as_soon_as_possible(**kwargs)
+        self.is_started = False
+
+    async def _run_task_as_soon_as_possible(self, force=False, **kwargs):
+        """
+        Wait until job _run() can be called
+        :param force: if True, force job task execution
+        """
+        if self._should_run_job(force=force):
+            await self._run(**kwargs)
+        else:
+            # wait for job dependencies to stop running
+            # and also this job to stop running
+            try:
+                events_to_wait = [
+                    asyncio.wait_for(
+                        dependency.running_task_event.wait(),
+                        self.DEPENDENCIES_WAIT_TIMEOUT,
+                    )
+                    for dependency in self.job_dependencies
+                ]
+                if not self.enable_multiple_runs:
+                    events_to_wait.append(
+                        asyncio.wait_for(
+                            self.running_task_event.wait(),
+                            self.SELF_RUNNING_WAIT_TIMEOUT,
+                        )
+                    )
+                await asyncio.gather(
+                    *events_to_wait,
+                )
+            except TimeoutError:
+                pass
+            finally:
+                await self._run(**kwargs)
+
+    def is_job_running(self):
+        """
+        :return: publicly is_running attribute value
+        """
+        return self.running_task_event.is_set()
+
+    def add_job_dependency(self, job):
+        """
+        Add a new job dependency
+        :param job: the new job dependency
+        """
+        self.job_dependencies.append(job)
+
+    async def _run(self, **kwargs):
+        """
+        Execute the job callback
+        Reset the last_execution_time
+        """
+        # Set to be able to await the event
+        self.running_task_event.set()
+        try:
+            await self.callback(**kwargs)
+        except Exception as exception:
+            self.logger.error(f"Failed to run job action : {exception}")
+        finally:
+            self.last_execution_time = time.time()
+            # Clear the event to trigger event waiters
+            self.running_task_event.clear()
```

Comment:
Are you sure about this set and clear ? To me it looks like they should be inverted: https://docs.python.org/3/library/asyncio-sync.html#asyncio.Event.set